diff --git a/qtwebengine/src/3rdparty/chromium/content/renderer/gpu/compositor_output_surface.cc b/qtwebengine/src/3rdparty/chromium/content/renderer/gpu/compositor_output_surface.cc
index 76998d6..0e80cef 100644
--- a/qtwebengine/src/3rdparty/chromium/content/renderer/gpu/compositor_output_surface.cc
+++ b/qtwebengine/src/3rdparty/chromium/content/renderer/gpu/compositor_output_surface.cc
@@ -86,9 +86,9 @@ bool CompositorOutputSurface::BindToClient(
   if (!context_provider()) {
     // Without a GPU context, the memory policy otherwise wouldn't be set.
     client->SetMemoryPolicy(cc::ManagedMemoryPolicy(
-        128 * 1024 * 1024,
+        256 * 1024 * 1024,
         gpu::MemoryAllocation::CUTOFF_ALLOW_NICE_TO_HAVE,
-        base::SharedMemory::GetHandleLimit() / 3));
+        cc::ManagedMemoryPolicy::kDefaultNumResourcesLimit));
   }
 
   return true;
